File: caracal/sdk/mandates.py
"""
Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
Caracal, a product of Garudex Labs

SDK Mandate Operations.

Provides mandate lifecycle management within a scoped context.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional

from caracal.logging_config import get_logger
from caracal.sdk.adapters.base import SDKRequest

if TYPE_CHECKING:
    from caracal.sdk.context import ScopeContext

logger = get_logger(__name__)


class MandateOperations:
    """Mandate lifecycle management within a scoped context.

    All methods inject scope headers and fire lifecycle hooks.
    """

    def __init__(self, scope: ScopeContext) -> None:
        self._scope = scope

    def _build_request(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
    ) -> SDKRequest:
        headers = dict(self._scope.scope_headers())
        return SDKRequest(
            method=method, path=path, headers=headers, body=body, params=params
        )

    async def _execute(self, request: SDKRequest) -> Any:
        scope_ref = self._scope.to_scope_ref()
        request = self._scope._hooks.fire_before_request(request, scope_ref)
        try:
            response = await self._scope._adapter.send(request)
            self._scope._hooks.fire_after_response(response, scope_ref)
            return response.body
        except Exception as exc:
            self._scope._hooks.fire_error(exc)
            raise

    # -- Public API --------------------------------------------------------

    async def create(
        self,
        agent_id: str,
        allowed_operations: List[str],
        expires_in: int,
        intent: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Create a new execution mandate."""
        body: Dict[str, Any] = {
            "agent_id": agent_id,
            "allowed_operations": allowed_operations,
            "expires_in": expires_in,
        }
        if intent:
            body["intent"] = intent
        if metadata:
            body["metadata"] = metadata
        req = self._build_request("POST", "/mandates", body=body)
        return await self._execute(req)

    async def validate(
        self,
        mandate_id: str,
        requested_action: str,
        requested_resource: str,
    ) -> Dict[str, Any]:
        """Validate a mandate for a specific action."""
        body = {
            "requested_action": requested_action,
            "requested_resource": requested_resource,
        }
        req = self._build_request(
            "POST", f"/mandates/{mandate_id}/validate", body=body
        )
        return await self._execute(req)

    async def revoke(
        self,
        mandate_id: str,
        revoker_id: str,
        reason: str,
        cascade: bool = True,
    ) -> Dict[str, Any]:
        """Revoke an execution mandate."""
        body = {
            "revoker_id": revoker_id,
            "reason": reason,
            "cascade": cascade,
        }
        req = self._build_request(
            "POST", f"/mandates/{mandate_id}/revoke", body=body
        )
        return await self._execute(req)

    async def get(self, mandate_id: str) -> Dict[str, Any]:
        """Get a mandate by ID."""
        req = self._build_request("GET", f"/mandates/{mandate_id}")
        return await self._execute(req)

    async def list(
        self,
        agent_id: Optional[str] = None,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """List mandates, optionally filtered by agent."""
        params: Dict[str, Any] = {"limit": limit}
        if agent_id:
            params["agent_id"] = agent_id
        req = self._build_request("GET", "/mandates", params=params)
        return await self._execute(req)

File: caracal-node/src/mandates.ts
/**
 * Copyright (C) 2026 Garudex Labs. All Rights Reserved.
 * Caracal, a product of Garudex Labs
 *
 * SDK Mandate Operations.
 */

import { SDKRequest } from './adapters/base';
import { ScopeContext } from './context';

export class MandateOperations {
  constructor(private readonly scope: ScopeContext) {}

  private buildReq(method: string, path: string, body?: Record<string, unknown>, params?: Record<string, unknown>): SDKRequest {
    return { method, path, headers: { ...this.scope.scopeHeaders() }, body, params };
  }

  private async exec(req: SDKRequest): Promise<unknown> {
    const scoped = await this.scope._hooks.fireBeforeRequest(req, this.scope.toScopeRef());
    try {
      const res = await this.scope._adapter.send(scoped);
      this.scope._hooks.fireAfterResponse(res, this.scope.toScopeRef());
      return res.body;
    } catch (e) {
      this.scope._hooks.fireError(e as Error);
      throw e;
    }
  }

  async create(options: {
    agentId: string;
    allowedOperations: string[];
    expiresIn: number;
    intent?: Record<string, unknown>;
    metadata?: Record<string, unknown>;
  }): Promise<unknown> {
    const body: Record<string, unknown> = {
      agent_id: options.agentId,
      allowed_operations: options.allowedOperations,
      expires_in: options.expiresIn,
    };
    if (options.intent) body.intent = options.intent;
    if (options.metadata) body.metadata = options.metadata;
    return this.exec(this.buildReq('POST', '/mandates', body));
  }

  async validate(options: {
    mandateId: string;
    requestedAction: string;
    requestedResource: string;
  }): Promise<unknown> {
    return this.exec(this.buildReq('POST', `/mandates/${options.mandateId}/validate`, {
      requested_action: options.requestedAction,
      requested_resource: options.requestedResource,
    }));
  }

  async revoke(options: {
    mandateId: string;
    revokerId: string;
    reason: string;
    cascade?: boolean;
  }): Promise<unknown> {
    return this.exec(this.buildReq('POST', `/mandates/${options.mandateId}/revoke`, {
      revoker_id: options.revokerId,
      reason: options.reason,
      cascade: options.cascade ?? true,
    }));
  }

  async get(mandateId: string): Promise<unknown> {
    return this.exec(this.buildReq('GET', `/mandates/${mandateId}`));
  }

  async list(options?: { agentId?: string; limit?: number }): Promise<unknown> {
    const params: Record<string, unknown> = { limit: options?.limit ?? 100 };
    if (options?.agentId) params.agent_id = options.agentId;
    return this.exec(this.buildReq('GET', '/mandates', undefined, params));
  }
}
